File: src/commands/rename-path.ts
/* 给路由文件打标记, 把标记打到最后,因为头部已经给了注释 */
import fs from 'fs-extra'
import type { ItemType } from '../types'
import {
  readFile
  //  writeFile
} from 'fs/promises'
import path from 'path'
import { createConsola } from 'consola'
import { getDependencies } from '../utils/router-utils'
import { getImportName } from './change-path'
const logger = createConsola({
  level: 4
})
const rootPath = process.cwd().replace(/\\/g, '/')
/**
 * 将单个字符串的首字母小写
 * @param str 字符串
 */
function fistLetterLower(str: string | String) {
  return str.charAt(0).toLowerCase() + str.slice(1)
}

export function toKebabCase(str: string) {
  const regex = /[A-Z]/g
  return fistLetterLower(str).replace(regex, (word: string) => {
    return '-' + word.toLowerCase()
  })
}

export function toCameCase(name: string) {
  // 使用正则表达式匹配中划线和随后的字符，同时将它们转换为大写
  let formattedName = name.replace(/-([a-z])/g, (_match, letter) => letter.toUpperCase())

  // 如果第一个字母不是大写，创建一个新的字符串并将其转换为大写
  if (formattedName[0] === formattedName[0].toLowerCase()) {
    formattedName = formattedName.charAt(0).toUpperCase() + formattedName.slice(1)
  }

  return formattedName
}
/**
 * 检测驼峰文件名
 * @param fileName 文件名
 */
export function checkCamelFile(fileName: string) {
  return /([a-z])([A-Z])/.test(fileName) || /([A-Z])/.test(fileName)
}

/**
 * 检测大驼峰文件名
 * @param fileName 文件名
 */
export function checkUperCamelFile(fileName: string) {
  return /([A-Z])/.test(fileName)
}

/**
 * @desc: 循环node, 改文件夹, 并把import 里面不合格的命名改合格
 */
export async function renameFoldPath(nodes: ItemType[], isCamelCase?: Boolean) {
  async function getNode(cpNodes: ItemType[]) {
    for (let index = 0; index < cpNodes.length; index++) {
      const ele = cpNodes[index]
      isCamelCase ? await renameFold(ele, true) : await renameFold(ele) // 下面已递归
      if (ele.children) {
        // 递归
        await getNode(ele.children)
      }
    }
  }
  await getNode(nodes)
}

/**
 * @desc: 循环node, 改文件, 改依赖, 思路:循环每个文件, 并把import 里面不合格的命名改合格
 */
export async function renameFilePath(nodes: ItemType[]) {
  async function getNode(cpNodes: ItemType[]) {
    for (let index = 0; index < cpNodes.length; index++) {
      const ele = cpNodes[index]
      if (ele.children) {
        // 递归
        await getNode(ele.children)
      } else {
        // 重命名文件
        await renameFile(ele)
        // 重写文件的import
        await rewriteFile(ele)
      }
    }
  }
  await getNode(nodes)
}

export async function renameCamelCaseFilePath(nodes: ItemType[]) {
  async function getNode(cpNodes: ItemType[]) {
    for (let index = 0; index < cpNodes.length; index++) {
      const ele = cpNodes[index]
      if (ele.children) {
        // 递归
        await getNode(ele.children)
      } else {
        // 重命名文件
        await renameCamelCaseFile(ele)
        // 重写文件的import
        await rewriteFile(ele, true)
      }
    }
  }
  await getNode(nodes)
}

async function rewriteFile(node: ItemType, isCamelCase?: Boolean) {
  let writeFlag = false
  try {
    const fileContent = await readFile(node.fullPath, 'utf-8')
    const lines = fileContent.split(/\n/g)

    const packageJsonPath = path.join(rootPath, 'package.json')
    const dependencies = await getDependencies(packageJsonPath)

    for (let index = 0; index < lines.length; index++) {
      const importLine = lines[index]
      if (importLine.includes('from')) {
        const importModuleName = getImportName(importLine, dependencies)

        if (isCamelCase) {
          if (checkUperCamelFile(importModuleName)) {
            const newName = toCameCase(path.parse(importModuleName).name)
            const [beforeFrom, afterFrom] = importLine.split('from')
            lines[index] = `${beforeFrom}from${afterFrom.replace(importModuleName, newName)}`
            writeFlag = true
          }
        } else if (checkCamelFile(importModuleName)) {
          const newName = toKebabCase(path.parse(importModuleName).name)
          const [beforeFrom, afterFrom] = importLine.split('from')
          lines[index] = `${beforeFrom}from${afterFrom.replace(importModuleName, newName)}`
          writeFlag = true
        }
      }
    }

    if (writeFlag) {
      const updatedFileContent = lines.join('\n')
      try {
        await fs.writeFile(node.fullPath, updatedFileContent, { encoding: 'utf8' })
        logger.success(`Rewrote file successfully: ${node.fullPath}`)
      } catch (writeError) {
        logger.error(`Failed to write file: ${node.fullPath}`, writeError)
      }
    }
  } catch (readError) {
    logger.error(`Failed to read file: ${node.fullPath}`, readError)
  }
}

/**
 * @desc: 重命名文件夹
 * @param {ItemType} node
 */
export async function renameFold(node: ItemType, isCamelCase?: boolean) {
  const filename = path.parse(node.fullPath).base

  const shouldRename = isCamelCase ? checkUperCamelFile(filename) : checkCamelFile(filename)
  if (shouldRename && node.isDir) {
    const obj = await replaceName(node.fullPath, isCamelCase)
    changePathFold(node, obj)
  }
}

/**
 * @desc: 重命名后, 子文件都会存在路径的更改,也就要递归处理(既可以处理文件夹, 也可以处理文件)
 */
export function changePathFold(node: ItemType, renameInfo: { newName: string; filename: string }): void {
  const { newName, filename } = renameInfo

  // If the node has children, recursively call this function on each child.
  if (node.children) {
    for (const childNode of node.children as ItemType[]) {
      changePathFold(childNode, renameInfo)
    }
  }

  // Update the full path and name of the current node.
  node.fullPath = node.fullPath.replace(filename, newName)
  node.name = node.name.replace(filename, newName)

  // Optionally, log once at the outermost call instead of in every recursion.
  // This can be controlled by a flag or condition check if needed.
  // if (isOutermostCall) {
  //   logger.info(node.fullPath, newName);
  // }
}

/**
 * @desc: 递归改所有路径名字
 * @param {ItemType} node
 * @param {object} obj
 */
export function changePathName(node: ItemType, obj: { newName: string; filename: string }, isCamelCase?: Boolean) {
  const { newName, filename } = obj
  if (node.fullPath.indexOf(filename) > -1) {
    if (node.imports.length > 0) {
      // import也要变化, 否则也会找不到路径
      const array = node.imports
      for (let j = 0; j < array.length; j++) {
        const ele = array[j]
        logger.info('import-ele: ', ele)
        array[j] = isCamelCase ? toCameCase(filename) : toKebabCase(ele)
        logger.info('更换import: ', array[j])
      }
    }
    node.fullPath = node.fullPath.replace(filename, newName)
    node.name = node.name.replace(filename, newName)
    logger.success('替换后的 node.fullPath:', node.fullPath)
  }
}

/**
 * @desc: 重命名文件
 * @param {ItemType} node
 */
export async function renameFile(node: ItemType) {
  const filename = path.parse(node.fullPath).base
  if (checkCamelFile(filename)) {
    const suffix = ['.js', '.vue', '.tsx'] // 这里只重命名js和vue文件
    const lastName = path.extname(node.fullPath)
    const flag = suffix.some((item) => lastName === item)
    if (flag) {
      const obj = await replaceName(node.fullPath)
      // 这里一定要更新node,否则后面找不到路径
      changePathName(node, obj)
    }
  }
}

export async function renameCamelCaseFile(node: ItemType) {
  const filename = path.parse(node.fullPath).base
  if (!checkUperCamelFile(filename)) {
    const suffix = ['.vue'] // 这里只重命名vue文件为大驼峰
    const lastName = path.extname(node.fullPath)
    const flag = suffix.some((item) => lastName === item)
    if (flag) {
      const obj = await replaceName(node.fullPath, true)
      // 这里一定要更新node,否则后面找不到路径
      changePathName(node, obj, true)
    }
  }
}

/**
 * 重命名文件夹 CamelCase || PascalCase => kebab-case
 * @param fullPath '/path/to/myFile.txt'
 * @return {newName:'my-file.txt','myFile.txt'}
 */
export async function replaceName(fullPath: string, isCamelCase?: Boolean) {
  const filename = path.parse(fullPath).base
  const newName = isCamelCase ? toCameCase(filename) : toKebabCase(filename)

  try {
    const oldPath = fullPath
    console.log('oldPath: ', oldPath)
    const newPath = oldPath.replace(filename, newName)
    const lastName = path.extname(newPath)
    if (!lastName) {
      // 处理目录
      if (fs.existsSync(newPath)) {
        await fs.copy(oldPath, newPath)
        await fs.rm(oldPath, { recursive: true }) // 删除目录
        return { newName, filename }
      }
    }
    // 处理文件
    if (await fs.pathExists(oldPath)) {
      await fs.rename(oldPath, newPath)
      logger.success(`${oldPath} renamed to: ${newPath}`)
    } else {
      logger.error(`File ${oldPath} does not exist.`)
    }
    logger.info(`${filename} is renamed done`)
    return { newName, filename }
  } catch (error) {
    logger.error(`Error renaming file/directory: ${error}`)
    throw error
  }
}

File: src/rename_path.rs
use crate::types::{FileNode, RenameInfo};
use crate::utils::{check_camel_file, check_upper_camel_file, to_camel_case, to_kebab_case, get_dependencies, get_import_name};
use anyhow::Result;
use std::path::Path;

fn rename_fold_path_sync(nodes: &mut [FileNode], is_camel_case: bool, dry_run: bool) -> Result<()> {
    for ele in nodes.iter_mut() {
        if ele.is_dir {
            if let Err(e) = rename_fold_sync(ele, is_camel_case, dry_run) {
                eprintln!("Warning: failed to rename folder {}: {}", ele.full_path, e);
            }
        }
        if let Some(children) = ele.children.as_mut() {
            rename_fold_path_sync(children, is_camel_case, dry_run)?;
        }
    }
    Ok(())
}

/// 递归重命名文件夹
pub async fn rename_fold_path(nodes: &mut [FileNode], is_camel_case: bool, dry_run: bool) -> Result<()> {
    tokio::task::block_in_place(|| rename_fold_path_sync(nodes, is_camel_case, dry_run))
}

fn rename_file_path_sync(nodes: &mut [FileNode], root_path: &Path, dry_run: bool) -> Result<()> {
    for ele in nodes.iter_mut() {
        if let Some(children) = ele.children.as_mut() {
            rename_file_path_sync(children, root_path, dry_run)?;
        } else {
            if let Err(e) = rename_file_sync(ele, dry_run) {
                eprintln!("Warning: failed to rename file {}: {}", ele.full_path, e);
            }
            if let Err(e) = rewrite_file_sync(ele, root_path, false, dry_run) {
                eprintln!("Warning: failed to rewrite file {}: {}", ele.full_path, e);
            }
        }
    }
    Ok(())
}

/// 递归重命名文件（kebab-case）
pub async fn rename_file_path(nodes: &mut [FileNode], root_path: &Path, dry_run: bool) -> Result<()> {
    tokio::task::block_in_place(|| rename_file_path_sync(nodes, root_path, dry_run))
}

fn rename_camel_case_file_path_sync(nodes: &mut [FileNode], root_path: &Path, dry_run: bool) -> Result<()> {
    for ele in nodes.iter_mut() {
        if let Some(children) = ele.children.as_mut() {
            rename_camel_case_file_path_sync(children, root_path, dry_run)?;
        } else {
            if let Err(e) = rename_camel_case_file_sync(ele, dry_run) {
                eprintln!("Warning: failed to rename file {}: {}", ele.full_path, e);
            }
            if let Err(e) = rewrite_file_sync(ele, root_path, true, dry_run) {
                eprintln!("Warning: failed to rewrite file {}: {}", ele.full_path, e);
            }
        }
    }
    Ok(())
}

/// 递归重命名文件（CamelCase）
pub async fn rename_camel_case_file_path(nodes: &mut [FileNode], root_path: &Path, dry_run: bool) -> Result<()> {
    tokio::task::block_in_place(|| rename_camel_case_file_path_sync(nodes, root_path, dry_run))
}

/// 重写文件中的 import 路径
fn rewrite_file_sync(node: &mut FileNode, root_path: &Path, is_camel_case: bool, dry_run: bool) -> Result<()> {
    let mut write_flag = false;
    let file_content = std::fs::read_to_string(&node.full_path)?;
    let mut lines: Vec<String> = file_content.lines().map(|s| s.to_string()).collect();

    let package_json_path = root_path.join("package.json");
    let dependencies = get_dependencies(&package_json_path);

    for line in lines.iter_mut() {
        if line.contains("from") {
            if let Some(import_module_name) = get_import_name(line, &dependencies) {
                let name = std::path::Path::new(&import_module_name)
                    .file_stem()
                    .and_then(|s| s.to_str())
                    .unwrap_or(&import_module_name);

                if is_camel_case {
                    if check_upper_camel_file(name) {
                        let new_name = to_camel_case(name);
                        let parts: Vec<&str> = line.split("from").collect();
                        if parts.len() == 2 {
                            *line = format!("{}from{}", parts[0], parts[1].replace(name, &new_name));
                            write_flag = true;
                        }
                    }
                } else if check_camel_file(name) {
                    let new_name = to_kebab_case(name);
                    let parts: Vec<&str> = line.split("from").collect();
                    if parts.len() == 2 {
                        *line = format!("{}from{}", parts[0], parts[1].replace(name, &new_name));
                        write_flag = true;
                    }
                }
            }
        }
    }

    if write_flag {
        let updated = lines.join("\n");
        if dry_run {
            println!("Dry-run: would rewrite file {}", node.full_path);
        } else {
            std::fs::write(&node.full_path, updated)?;
            println!("Rewrote file successfully: {}", node.full_path);
        }
    }

    Ok(())
}

/// 判断两个路径在大小写不敏感情况下是否指向同一位置（用于 Windows）
fn is_same_path_ignore_case(a: &Path, b: &Path) -> bool {
    let a_components: Vec<_> = a.components().collect();
    let b_components: Vec<_> = b.components().collect();
    if a_components.len() != b_components.len() {
        return false;
    }
    a_components.iter().zip(b_components.iter()).all(|(ac, bc)| {
        ac.as_os_str().to_string_lossy().to_lowercase()
            == bc.as_os_str().to_string_lossy().to_lowercase()
    })
}

/// 重命名单个文件夹
fn rename_fold_sync(node: &mut FileNode, is_camel_case: bool, dry_run: bool) -> Result<()> {
    let filename = std::path::Path::new(&node.full_path)
        .file_name()
        .and_then(|s| s.to_str())
        .unwrap_or(&node.name);

    let should_rename = if is_camel_case {
        check_upper_camel_file(filename)
    } else {
        check_camel_file(filename)
    };

    if should_rename && node.is_dir {
        let info = replace_name_sync(&node.full_path, is_camel_case, dry_run)?;
        // 只有当重命名实际执行了（old_name != new_name）才更新内存路径
        // dry-run 模式下也不更新，避免后续节点路径与物理路径不一致
        if !dry_run && info.old_name != info.new_name {
            change_path_fold(node, &info);
        }
    }
    Ok(())
}

/// 只替换路径中最后一个匹配的组件名
fn replace_last_path_component(path: &str, old_name: &str, new_name: &str) -> String {
    if let Some(pos) = path.rfind(&format!("/{}", old_name)) {
        let after = &path[pos + 1 + old_name.len()..];
        if after.is_empty() || after.starts_with('/') {
            return format!("{}{}{}", &path[..pos + 1], new_name, after);
        }
    }
    if path == old_name {
        return new_name.to_string();
    }
    path.to_string()
}

/// 递归更新后代节点路径：用 new_prefix 替换以 old_prefix 开头的路径
fn update_descendant_path(node: &mut FileNode, old_prefix: &str, new_prefix: &str) {
    if node.full_path == old_prefix {
        node.full_path = new_prefix.to_string();
    } else if node.full_path.starts_with(&format!("{}/", old_prefix)) {
        node.full_path = format!("{}{}", new_prefix, &node.full_path[old_prefix.len()..]);
    }
    if let Some(children) = node.children.as_mut() {
        for child in children.iter_mut() {
            update_descendant_path(child, old_prefix, new_prefix);
        }
    }
}

/// 重命名后更新子节点路径
pub fn change_path_fold(node: &mut FileNode, rename_info: &RenameInfo) {
    // 更新当前节点
    if node.name == rename_info.old_name {
        node.name = rename_info.new_name.clone();
    }
    let old_full_path = node.full_path.clone();
    node.full_path = replace_last_path_component(&node.full_path, &rename_info.old_name, &rename_info.new_name);
    let new_full_path = node.full_path.clone();

    // 递归更新子节点（用前缀替换，避免同名组件误替换）
    if let Some(children) = node.children.as_mut() {
        for child in children.iter_mut() {
            update_descendant_path(child, &old_full_path, &new_full_path);
        }
    }
}

/// 重命名文件（kebab-case）
fn rename_file_sync(node: &mut FileNode, dry_run: bool) -> Result<()> {
    let filename = std::path::Path::new(&node.full_path)
        .file_name()
        .and_then(|s| s.to_str())
        .unwrap_or(&node.name);

    if check_camel_file(filename) {
        let suffixes = [".js", ".vue", ".tsx"];
        let last_name = std::path::Path::new(&node.full_path)
            .extension()
            .and_then(|s| s.to_str())
            .unwrap_or("");
        let flag = suffixes.iter().any(|item| *item == format!(".{}", last_name));
        if flag {
            let info = replace_name_sync(&node.full_path, false, dry_run)?;
            change_path_name(node, &info, false);
        }
    }
    Ok(())
}

/// 重命名文件（CamelCase）
fn rename_camel_case_file_sync(node: &mut FileNode, dry_run: bool) -> Result<()> {
    let filename = std::path::Path::new(&node.full_path)
        .file_name()
        .and_then(|s| s.to_str())
        .unwrap_or(&node.name);

    if !check_upper_camel_file(filename) {
        let suffixes = [".vue"];
        let last_name = std::path::Path::new(&node.full_path)
            .extension()
            .and_then(|s| s.to_str())
            .unwrap_or("");
        let flag = suffixes.iter().any(|item| *item == format!(".{}", last_name));
        if flag {
            let info = replace_name_sync(&node.full_path, true, dry_run)?;
            change_path_name(node, &info, true);
        }
    }
    Ok(())
}

/// 更新节点路径和 import
pub fn change_path_name(node: &mut FileNode, obj: &RenameInfo, is_camel_case: bool) {
    if node.full_path.contains(&obj.old_name) {
        if !node.imports.is_empty() {
            for ele in node.imports.iter_mut() {
                *ele = if is_camel_case {
                    to_camel_case(&obj.old_name)
                } else {
                    to_kebab_case(ele)
                };
            }
        }
        node.full_path = node.full_path.replace(&obj.old_name, &obj.new_name);
        node.name = node.name.replace(&obj.old_name, &obj.new_name);
    }
}

/// 重命名文件或文件夹
fn replace_name_sync(full_path: &str, is_camel_case: bool, dry_run: bool) -> Result<RenameInfo> {
    let path_obj = std::path::Path::new(full_path);
    let filename = path_obj.file_name().and_then(|s| s.to_str()).unwrap_or("");
    let new_name = if is_camel_case {
        to_camel_case(filename)
    } else {
        to_kebab_case(filename)
    };

    let old_path = std::path::Path::new(full_path);
    let new_path = old_path.with_file_name(&new_name);

    // 如果新旧路径在大小写不敏感情况下相同（如 Windows 上的 Home -> home），直接跳过
    if is_same_path_ignore_case(old_path, &new_path) {
        println!("Skipping rename: {} and {} are the same path (case-insensitive)", old_path.display(), new_path.display());
        return Ok(RenameInfo {
            new_name: filename.to_string(),
            old_name: filename.to_string(),
        });
    }

    if old_path.is_dir() {
        if new_path.exists() {
            if dry_run {
                println!("Dry-run: would copy dir {} -> {} and remove {}", old_path.display(), new_path.display(), old_path.display());
            } else {
                copy_dir_all_sync(old_path, &new_path)?;
                std::fs::remove_dir_all(old_path)?;
            }
        } else {
            if dry_run {
                println!("Dry-run: would rename {} -> {}", old_path.display(), new_path.display());
            } else {
                match std::fs::rename(old_path, &new_path) {
                    Ok(_) => {
                        println!("{} renamed to: {}", old_path.display(), new_path.display());
                    }
                    Err(e) => {
                        eprintln!("Direct rename failed for {}: {}, trying copy-merge...", old_path.display(), e);
                        copy_dir_all_sync(old_path, &new_path)?;
                        if let Err(remove_err) = std::fs::remove_dir_all(old_path) {
                            eprintln!("Warning: could not remove old directory {} after copy-merge: {}", old_path.display(), remove_err);
                        }
                        println!("{} copy-merged to: {}", old_path.display(), new_path.display());
                    }
                }
            }
        }
    } else if old_path.exists() {
        if dry_run {
            println!("Dry-run: would rename {} -> {}", old_path.display(), new_path.display());
        } else {
            std::fs::rename(old_path, &new_path)?;
            println!("{} renamed to: {}", old_path.display(), new_path.display());
        }
    } else {
        eprintln!("File {} does not exist.", old_path.display());
    }

    println!("{} is renamed done", filename);
    Ok(RenameInfo {
        new_name,
        old_name: filename.to_string(),
    })
}

/// 递归复制目录
fn copy_dir_all_sync(src: &std::path::Path, dst: &std::path::Path) -> Result<()> {
    std::fs::create_dir_all(dst)?;
    let entries = std::fs::read_dir(src)?;
    for entry in entries {
        let entry = entry?;
        let entry_path = entry.path();
        let file_name = entry.file_name();
        let dest_path = dst.join(&file_name);
        if entry_path.is_dir() {
            copy_dir_all_sync(&entry_path, &dest_path)?;
        } else {
            std::fs::copy(&entry_path, &dest_path)?;
        }
    }
    Ok(())
}
